src/main/java/org/betonquest/betonquest/quest/event/give/GiveEvent.java

```diff
@@ -0,0 +1,140 @@
+     * @param backpack              whether to put the items to the player's backpack
+     */
+    public GiveEvent(final Item[] questItems, final NotificationSender itemsGivenSender, final NotificationSender itemsInBackpackSender, final NotificationSender itemsDroppedSender, final boolean backpack) {
+        this.questItems = Arrays.copyOf(questItems, questItems.length);
+        this.itemsGivenSender = itemsGivenSender;
+        this.itemsInBackpackSender = itemsInBackpackSender;
+        this.itemsDroppedSender = itemsDroppedSender;
+        this.backpack = backpack;
+    }
+
+    @Override
+    public void execute(final Profile profile) throws QuestRuntimeException {
+        final Player player = profile.getOnlineProfile().get().getPlayer();
+        Arrays.stream(questItems).toList().forEach(item -> {
+            final QuestItem questItem = item.getItem();
+            final int amount = item.getAmount().getInt(profile);
+            giveItems(profile, player, questItem, amount);
+            final String questItemName = questItem.getName() == null
+                    ? questItem.getMaterial().toString().toLowerCase(Locale.ROOT).replace("_", " ")
+                    : questItem.getName();
+            itemsGivenSender.sendNotification(profile, questItemName, String.valueOf(amount));
+        });
+    }
+
+    private void giveItems(final Profile profile, final Player player, final QuestItem questItem, final int totalAmount) {
+        int amount = totalAmount;
+        while (amount > 0) {
+            boolean fullInventory = false;
+            final ItemStack itemStackTemplate = questItem.generate(1, profile);
+            final int stackSize = Math.min(amount, itemStackTemplate.getMaxStackSize());
+            ItemStack itemStack = itemStackTemplate.clone();
+            itemStack.setAmount(stackSize);
+            if (!backpack) {
+                final ItemStack leftItems = giveToInventory(player, itemStack);
+                if (leftItems == null) {
+                    amount -= stackSize;
+                    continue;
+                } else {
+                    itemStack = leftItems;
+                    fullInventory = true;
+                }
+            }
+            if (Utils.isQuestItem(itemStack)) {
+                giveToBackpack(profile, itemStack);
+                if (fullInventory) {
+                    itemsInBackpackSender.sendNotification(profile);
+                }
+            } else {
+                dropItems(player, itemStack);
+                itemsDroppedSender.sendNotification(profile);
+            }
+            amount -= stackSize;
+        }
+    }
+
+    /**
+     * Gives the item to the player. Returns null if the item was given successfully, otherwise returns the
+     * items that was not given.
+     *
+     * @param player    the player to give the item to
+     * @param itemStack the item to give
+     * @return the item that was not given
```

Comment:
```suggestion
     * Gives the item stack to the player. Returns null if all items of the stack were given successfully,
     * otherwise returns the items that were not given.
     *
     * @param player    the player to give the item to
     * @param itemStack the items to give
     * @return the items that could not be given
```